### PR TITLE
Participant Composite

### PR DESCRIPTION
**Pull request by frostbyte73**

### CODE REVIEW

**Review by biglittlebigben — APPROVED** [review]
Great! I think we may just consider relying a bit more on contexts for some of the cancellation/timeout handling.

File: pkg/gstreamer/bin.go [hunk 1]
```diff
@@ -253,7 +257,19 @@ func (b *Bin) RemoveSinkBin(name string) (bool, error) {
 }
 
 func (b *Bin) SetState(state gst.State) error {
-	return b.bin.SetState(state)
+	stateErr := make(chan error, 1)
+	go func() {
+		stateErr <- b.bin.SetState(state)
+	}()
+	select {
+	case <-time.After(stateChangeTimeout):
```
[the remark this answers is not in the document]
Comment (frostbyte73) [reply pkg/gstreamer/bin.go:265]: if you're passing the context into other functions/doing more with deadlines I'd agree, but for simple cases like this I prefer the simplicity of time.After. One line less without the `defer cancel()`

[the remark this answers is not in the document]
Comment (biglittlebigben) [reply pkg/gstreamer/bin.go:265]: Indeed. My point would be that I think that ideal `(b *Bin) SetState` (and potentially callers as well) should take a context as well, allowing the caller to cancel the wait on timeout.

Retrofitting all this is beyond the scope of this PR, of course.

File: pkg/pipeline/source/sdk.go [hunk 2]
```diff
@@ -326,16 +380,21 @@ func (s *SDKSource) onTrackSubscribed(track *webrtc.TrackRemote, pub *lksdk.Remo
 	if !s.initialized.IsBroken() {
 		s.mu.Lock()
 		switch s.RequestType {
+		case types.RequestTypeParticipant:
+			s.filenameReplacements["{publisher_identity}"] = s.Identity
```
Comment (davidzhao) [inline pkg/pipeline/source/sdk.go:384]: would we make pID a replacement policy too? not sure if that needs to be onTrackSubscribed tho.

**Review by davidzhao — APPROVED** [review]
nice!

File: pkg/pipeline/source/sdk.go (hunk 2 above)
Comment (frostbyte73) [reply pkg/pipeline/source/sdk.go:384]: could add ID if/when someone asks, I think for the most part identity is more useful

Comment (davidzhao) [reply pkg/pipeline/source/sdk.go:384]: yeah you are right.. this is participant composite.. identity is the main thing people would care about.

**Pull request merged**